src/zkstar.rs

```diff
@@ -0,0 +1,118 @@
+use crate::errors::InternalError;
+use libpaillier::unknown_order::BigNumber;
+use serde::{Deserialize, Serialize};
+
+/// A warpper around the modulus of the group [ZStarN] is defined over
+/// This allows passing the modulus by reference and avoid serializing it
+#[derive(Serialize)]
+struct ZStarNBuilder {
+    N: BigNumber,
+}
+
+impl ZStarNBuilder {
+    pub fn new(n: BigNumber) -> Self {
+        ZStarNBuilder { N: n }
+    }
+    /// The only way to construct an element of [ZStarN] is to validate that [ZStarNUnverified]
+    /// is properly constructed with respect to the current [ZStarNBuilder]
+    pub fn validate(&self, unverified: ZStarNUnverified) -> Result<ZStarN, InternalError> {
+        if unverified.value().is_zero() {
+            return Err(InternalError::IsZero);
+        } else if unverified.value() > self.value() {
+            return Err(InternalError::LargerThanModulus);
+        }
+        let result = unverified.value().extended_gcd(self.value());
+        if result.gcd != BigNumber::one() {
+            return Err(InternalError::NotCoprime);
+        }
+        Ok(ZStarN {
+            value: unverified.value().clone(),
+            builder: self,
+        })
+    }
+    /// Return the modulus of [ZStarNBuilder]
+    fn value(&self) -> &BigNumber {
+        &self.N
+    }
+}
+
+/// An verifier element of [ZStarN]
+#[derive(Serialize)]
+struct ZStarN<'a> {
+    value: BigNumber,
+    builder: &'a ZStarNBuilder,
+}
+
+impl<'a> ZStarN<'a> {
```

Comment:
todo: We also need a function to generate a random element and replace `utils::random_bn_in_z_star`. I would suggest keeping the utils function for now (so you don't have to go change all the use cases); we'll plan to remove it in the second PR.

I'm flexible about whether you prefer `ZStarN::random() -> Self` or `ZStarNBuilder::random() -> ZStarN`.

Reply:
I prefer `ZStarN::random() -> Self`, because I am guessing that that's what the programmer will intend. The later one can be a bit confusing if only because the type of what its called on does not match the type that is returned :) Let me know if that sounds reasonable.

Reply:
Regarding the method itself: are we still thinking of NOT passing the rng ? Its usually cleaner to pass it because it makes testing and modifying it easier.

Reply:
Yeah, I like that too.